AuthZ 2.0 - [Breaking] AuthorizeResult/Failure

HaoK opened the pull request:
@blowdart @ajcvickers @Tratcher 

Needed for: https://github.com/aspnet/Security/issues/901

First step needed to surface why authorization fails, returning a result which can contain failure information instead of just a bool.

Also added IAuthorizationHandlerProvider, and reworked IAuthorizationEvaluator to work with the new API.

Fixes: https://github.com/aspnet/Security/issues/1176

dnfclas commented:
@HaoK,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

HaoK commented on src/Microsoft.AspNetCore.Authorization.Policy/Microsoft.AspNetCore.Authorization.Policy.csproj:
```diff
@@ -0,0 +1,30 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore.Authentication.Abstractions" Version="$(AspNetCoreVersion)" />
```
AuthZ currently doesn't depend on Http or AuthN, if we are ok taking these dependencies in the main AuthZ package, we don't need to split out the authorize helper logic

PinpointTownes commented on src/Microsoft.AspNetCore.Authorization/AuthorizeFailure.cs:
```diff
@@ -0,0 +1,46 @@
+    /// <summary>
+    /// Encapsulates a failure result of <see cref="IAuthorizationService.AuthorizeAsync(ClaimsPrincipal, object, IEnumerable{IAuthorizationRequirement})"/>.
+    /// </summary>
+    public class AuthorizeFailure
```
All these `verb` + `Result` classes sound a bit weird. Why not `AuthorizationFailure`?

HaoK replied:
I started with that, but we went with AuthenticateResult (there that was meant to signify that it was a result for the Authenticate method).  I could go either way here since I've had it both ways in this code already :)

HaoK commented on src/Microsoft.AspNetCore.Authorization.Policy/AuthorizationPolicyResult.cs:
```diff
@@ -0,0 +1,35 @@
+        /// <summary>
+        /// If true, means the callee should challenge and try again.
+        /// </summary>
+        public bool Challenged { get; private set; }
```
Challenged => RequiresAuthentication?

HaoK commented:
Updated the PolicyEvaluator to have the new logic where when authorization fails, forbid is returned if any schemes were successfully authenticated.

Tratcher commented on src/Microsoft.AspNetCore.Authorization.Policy/PolicyEvaluator.cs:
```diff
@@ -0,0 +1,89 @@
+                    }
+                }
+
+                if (newPrincipal == null)
```
!= null?

HaoK replied:
Doh yeah, looks like I am missing tests for the Authenticate method too :)

HaoK commented:
- Renamed AuthorizeResult/Failure => AuthorizationResult/Failure
- Eliminated the policy authentication result class trying to reuse AuthenticateResult directly instead.
- added tests for authentication of policy evaluator

HaoK commented on src/Microsoft.AspNetCore.Authorization.Policy/PolicyEvaluator.cs:
```diff
@@ -0,0 +1,89 @@
+                    return AuthenticateResult.None();
+                }
+            }
+            return AuthenticateResult.None();
```
We could explicitly check the result of AuthenticateAsync to determine if we want to return success/none instead to control the behavior of forbid/challenge if authorization fails later.

HaoK commented:
@Tratcher updated to react to the breaking auth changes in HttpAbstractions

HaoK commented:
e940cdb36bc7617af3d762ff9cc853a8fef55b0a
